modules/tensor_mechanics/src/materials/FiniteStrainUObasedCP.C

```diff
@@ -0,0 +1,851 @@
+
+  // Check for stress residual tolerance
+  while (rnorm > _rtol * rnorm0 && rnorm0 > _abs_tol && iter <  _maxiter)
+  {
+    // Calculate stress increment
+    dpk2 = - jac.invSymm() * resid;
+    _pk2[_qp] = _pk2[_qp] + dpk2;
+    calcResidJacob(resid, jac);
+    // update _fp_prev_inv
+    internalVariableUpdateNRiteration();
+
+    if (_err_tol)
+    {
+#ifdef DEBUG
+      mooseWarning("FiniteStrainUObasedCP: Slip increment exceeds tolerance - Element number " << _current_elem->id() << " Gauss point = " << _qp);
+#endif
+      return;
+    }
+
+    rnorm_prev = rnorm;
+    rnorm = resid.L2norm();
+
+    if (_use_line_search && rnorm > rnorm_prev && !lineSearchUpdate(rnorm_prev, dpk2))
+    {
+#ifdef DEBUG
+      mooseWarning("FiniteStrainUObasedCP: Failed with line search");
+#endif
+      _err_tol = true;
+      return;
+    }
+
+    if (_use_line_search)
+      rnorm = resid.L2norm();
+
+    iter++;
+  }
+
+  if (iter >= _maxiter)
+  {
+#ifdef DEBUG
+    mooseWarning("FiniteStrainUObasedCP: Stress Integration error rmax = " << rnorm);
+#endif
+    _err_tol = true;
+  }
+}
+
+void
+FiniteStrainUObasedCP::postSolveStress()
+{
+  if (_max_substep_iter == 1)
+    // No substepping
+    _fp[_qp] = _fp_inv.inverse();
+  else
+  {
+    if (_last_step_iter)
+      _fp[_qp] = _fp_inv.inverse();
+    else
+    {
+      _fp_old_inv = _fp_inv;
+      _pk2_tmp_old = _pk2[_qp];
+    }
+  }
+}
+
+// Update slip system resistance. Overide to incorporate new slip system resistance laws
+void
+FiniteStrainUObasedCP::updateSlipSystemResistance()
+{
+  for (unsigned int i = 0; i < _num_uo_state_vars; ++i)
+    _state_vars_prev[i] = (*_mat_prop_state_vars[i])[_qp];
+
+  for (unsigned int i = 0; i < _num_uo_state_var_evol_rate_comps; ++i)
+    _uo_state_var_evol_rate_comps[i]->calcStateVariableEvolutionRateComponent(_qp, (*_mat_prop_state_var_evol_rate_comps[i])[_qp]);
+
+  for (unsigned int i = 0; i < _num_uo_state_vars; ++i)
+  {
+    if (!_uo_state_vars[i]->updateStateVariable(_qp, _dt, (*_mat_prop_state_vars[i])[_qp]))
+      _err_tol = true;
+  }
+
+  for (unsigned int i = 0; i < _num_uo_slip_rates; ++i)
+    _uo_slip_resistances[i]->calcSlipResistance(_qp, (*_mat_prop_slip_resistances[i])[_qp]);
+}
+
+// Calculates stress residual equation and jacobian
+void
+FiniteStrainUObasedCP::calcResidJacob(RankTwoTensor & resid, RankFourTensor & jac)
+{
+  calcResidual(resid);
+  if (_err_tol)
+    return;
+  calcJacobian(jac);
+}
+
+void
+FiniteStrainUObasedCP::getSlipIncrements()
+{
+  for (unsigned int i = 0; i < _num_uo_slip_rates; ++i)
+  {
+    if (!_uo_slip_rates[i]->calcSlipRate(_qp, _dt, (*_mat_prop_slip_rates[i])[_qp]))
+    {
+      _err_tol = true;
+      return;
+    }
+  }
+}
+
+void
+FiniteStrainUObasedCP::calcResidual(RankTwoTensor &resid)
+{
+  RankTwoTensor iden, ce, ee, ce_pk2, eqv_slip_incr, pk2_new;
+
+  iden.zero();
+  iden.addIa(1.0);
+
+  eqv_slip_incr.zero();
+
+  // _fp_inv  ==> _fp_prev_inv
+  _fe = _dfgrd_tmp * _fp_prev_inv;
```

Comment:
Is This fe used for any calculation? You may like to remove this. Even Fp_prev_inv may be removed since tau is just based on pk2.